source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opt/scalar_analysis.h"
+#include <algorithm>
+#include <functional>
+#include <string>
+#include <utility>
+
+#include "opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+SENode* ScalarEvolutionAnalysis::CreateNegation(SENode* operand) {
```

Comment:
Surely not all SENodes can be negated? (e.g. ValueUnknown or CanNotCompute).  Assert for unsupported nodes?

Reply:
We allow ValueUnknowns to be negated because we can still reason that X - X is zero even if we don't know the value